src/pudl/output/ferc1.py

```diff
@@ -1337,3 +1339,93 @@ def remove_intra_table_calculated_values(exploded: pd.DataFrame) -> pd.DataFrame
+# XBRL Calculation Tree
+################################################################################
+Ferc1XbrlCalculationComponent = ForwardRef("Ferc1XbrlCalculationComponent")
+Ferc1XbrlCalculation = ForwardRef("Ferc1XbrlCalculation")
+
+
+class Ferc1XbrlCalculationComponent(BaseModel):
+    """A class representing the components of a FERC 1 XBRL Metadata calculation."""
+
+    name: str
+    weight: float
+    source_tables: list[str]
+    calculation: Ferc1XbrlCalculation | None = None
+
+    @validator("source_tables")
+    def has_unique_source_table(cls, value):
+        """Ensure that each calculation component has a unique source table."""
+        if len(value) == 1:
+            return value
+        else:
+            raise ValueError(
+                "Exploding tables requires calculations to have unique source tables."
+            )
+
+
+class Ferc1XbrlCalculation(BaseModel):
+    """A class representing a calculation from the FERC 1 XBRL Metadata."""
+
+    components: list[Ferc1XbrlCalculationComponent] = []
+    source_table: str
+    xbrl_factoid: str
+    xbrl_factoid_original: str
+
+    @property
+    def referenced_tables(self: Self) -> set[str]:
+        """Identify the set of all referenced tables in a calculation."""
+        return {c for x in self.calculations for c in x.source_tables}
+
+    def is_inter_table(self: Self, source_table: str | None = None) -> bool:
+        """Determine whether a calculation involves more than one table."""
+        all_tables = self.referenced_tables
+        if source_table is not None:
+            all_tables = all_tables.union({source_table})
+        return True if len(all_tables) > 1 else False
+
+
+Ferc1XbrlCalculationComponent.update_forward_refs()
+Ferc1XbrlCalculation.update_forward_refs()
+
+
+def resolve_calculation_components(
+    exploded_meta: pd.DataFrame,
+    table_name: str,
+    xbrl_factoid: str,
+) -> Ferc1XbrlCalculation:
+    """Build a calculation given exploded metadata, xbrl_factoid, and table_name."""
+    idx_cols = ["table_name", "xbrl_factoid"]
+    if exploded_meta.index.names != idx_cols:
+        raise AssertionError(
+            f"Exploded metadata must be indexed by {idx_cols}, but found "
+            f"{exploded_meta.index.names=}"
+        )
+    if not exploded_meta.index.is_unique:
+        raise AssertionError("Found non-unique index in exploded metadata.")
+    idx = (table_name, xbrl_factoid)
+    components = json.loads(exploded_meta.at[idx, "calculations"])
+    components = [Ferc1XbrlCalculationComponent(**c) for c in components]
+    new_components = []
+    for c in components:
+        new_comp = Ferc1XbrlCalculationComponent(
+            name=c.name,
+            weight=c.weight,
+            source_tables=c.source_tables,
+            calculation=resolve_calculation_components(
+                exploded_meta=exploded_meta,
+                table_name=c.source_tables[0],
```

Comment:
If the JSON object representing the calculation has multiple source tables, is it safe to just ignore the tail of the array?

Reply:
The only reason this is okay is because we're requiring `source_tables` to be of length 1 elsewhere in the process.